docs: ensure doc consistancy and correctness (#87)
asyncapi/generator-model-sdk: docs/simplification.md

# Simplifying JSON Schema to CommonModel

The library simplify the JSON Schema from data validation rules to data definitions (`CommonModel`(s)). We do this because JSON Schema files can be extremely complex and can in many cases be simplified down to a bare minimum representation. Now keep in mind that there is a difference between validating input and the underlying structure of the data. In our case we do not care directly about the validation of input but rather how we display the underlying structure correctly.

This document is expected to be a supplement to the code to better understand how the simplifier works.

## Simplifier

> The main functionality is located in the `Simplifier` class. This class ensures to recursively create (or retrieve from a cache) a `CommonModel` representation of a Schema. We have tried to keep the the functionality split out into separate functions to reduce complexity and ensure it is easier to maintain. This main function also ensures to split any created models into separate ones if needed.

Reviewer: Minor: Little typo there

```suggestion
The main functionality is located in the `Simplifier` class. This class ensures to recursively create (or retrieve from a cache) a `CommonModel` representation of a Schema. We have tried to keep the functionality split out into separate functions to reduce complexity and ensure it is easier to maintain. This main function also ensures to split any created models into separate ones if needed.
```

- [Required](#determining-the-required-properties-for-the-model)
- [Extend](#determining-the-extend-for-the-model)

> <br/>

Reviewer: I'm not sure this line break provides any improvement here. Also, if needed, you could add another empty line here as per markdown standards.

## Determining the type for the model

> In order to determine all the possible types a model can be based on schema, we both infer and use existing definitions of types, however we need to define a precedence for JSON Schema keywords for in which order they types are inferred or determined.

Reviewer: I think I understand what this sentence says, but took me a while to figure out. Would something like the following still mean the same you wanted to say?

```suggestion
In order to determine all the possible schema types a model can be based on, we both infer and use existing definitions of types. 
Precedence for JSON Schema keywords matters, meaning we need to evaluate them in a particular order.
```

Author: @smoya maybe the following is better?
```suggestion
In order to determine all the possible types for a model, we both infer and use existing definitions of types. 
Precedence for JSON Schema keywords needs to be predetermined to ensure consistency.
```

Reviewer: @jonaslagoni Yup, looks really good to me 👍

**Precedence of JSON Schema keywords**

> The precedence of keywords are in which order we infer or determine types in. Notice it goes form left to right meaning first we set the type using the `type` keyword and then it moves right applying them one by one. The following are the precedence for determining types:

Reviewer: Most of this would be already mentioned in https://github.com/asyncapi/generator-model-sdk/pull/87/files#r580499801 so I think it can partially be removed.

> <p align="center">type --> allOf --> oneOf --> anyOf --> then --> else --> enum --> const --> not</p>

Reviewer: Totally subjective: I saw how this is rendered and not pretty convinced it improves readability in my case.

Author: The reason for the centering was that I felt like everything blended together otherwise 🤔

Reviewer: Fair enough. just felt a bit weird to me but it seems fine otherwise :) Thanks for your reasoning!

**Absence of data type**

If a schema is defined as `true` or `false` we infer all possible JSON Schema types.

**Enum and const**

`enum` and `const` are ONLY used to infer the type if it has not already been defined. Type is inferred from the provided value(s).

> **Not**

Reviewer: I think you can use an h3 here (`###` prefix) so it will render as you expect.

> if defined removes any defined/inferred types defined in `not`.

Reviewer: ```suggestion
If defined, removes any defined/inferred types defined in `not`.
```

<br/>

## Determining the enums for the model

> In order to determine the possible enums a model can be we both infer and use existing definitions, however we need to define a precedence for JSON Schema keywords and in which order they are applied.

Reviewer: Same as https://github.com/asyncapi/generator-model-sdk/pull/87/files#r580500164. I feel it sounds a bit redundant.
